### DISCUSSION TITLE
Information about parameters precision on Loihi

### DISCUSSION THREAD

**Discussion by fabrizio-ottati** [lava-nc/lava-dl#259] [Q&A]
Hi everyone. I am trying to estimate the model footprint of a network deployed to Loihi 1 from the literature. 
Given the CUBA-LIF implementation, I would like to know the precision of:
- voltage state.
- current state.
- the spiking threshold.
- the leakage parameters (are they shared among current and voltage?).
- weights.
- given a DNN, are the parameters (threshold and leakages) shared across a layer or in the entire network? Or is it possible to choose a different set of parameters per neuron (or per channel, in a CNN)?
- leakage (is it applied via bit-shift)?

Could you confirm these numbers or redirect me to the proper documentation? Thanks in advance!
Fabrizio

**Comment by Michaeljurado42** [comment]
I think I can be of assistance here because I asked [a very similar question](https://github.com/IntelLabs/IntelNeuromorphicDNSChallenge/issues/19) recently.

For voltage and current states you can use the [lava process model code](https://github.com/lava-nc/lava/blob/main/src/lava/proc/lif/models.py#L80) which lists the precisions for each field. For example, voltage and current are 24 bit signed integers. 

As for the weights, @bamsumit  calculates the Kilobytes used by each layer in [the issue I listed](https://github.com/IntelLabs/IntelNeuromorphicDNSChallenge/issues/19). If your connections are sparse then you can [reduce the memory footprint of the synapses](https://github.com/lava-nc/lava-dl/pull/238) however.

As for the leakage, that is also specified in the[ lava process model](https://github.com/lava-nc/lava/blob/main/src/lava/proc/lif/models.py#L149) code. It looks like to me that the bit shift is applied to scale down the voltage after applying a decay via an integer multiplication.

What I cannot answer is whether or not the neuron parameters are shared across a population or are defined on a per neuron basis when you port SNNs to Loihi2. However, from personal experience, the number of synapses in an SNN tends to be so much larger than the number of neurons and thus have a much more substantial memory footprint. 

_In bocca al lupo!_

**Reply by fabrizio-ottati** [reply]
Hi Michael, thank you very much :)
Actually, I do not intend on using Loihi, but I needed to estimate the model size to compare to my hardware implementation, so this information is super useful! In fact, in my CNN the weights dominate the memory footprint. _Grazie mille_ :)
